orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -104,3 +105,109 @@ def _create_iobinding(io_binding, inputs, model, device):
 
     for value_info in model.graph.output:
         io_binding.bind_output(value_info.name, device.type, device_id=get_device_index(device))
+
+def _load_aten_op_executor_cpp_extension(verbosity, is_rocm_pytorch):
+    aten_op_executor_cpp_source = """
+#include <torch/torch.h>
+#include <ATen/DLConvertor.h>
+#include <unordered_map>
+#include <tuple>
+#include <vector>
+
+class ATenOperatorCache {
+ public:
+  static ATenOperatorCache& Instance() {
+    static ATenOperatorCache instance;
+    return instance;
+  }
+
+  std::shared_ptr<torch::jit::Operator> GetOperator(const std::string& op_name) {
+    if (ops_.find(op_name) == ops_.end()) {
+      auto& ops = torch::jit::getAllOperatorsFor(torch::jit::Symbol::fromQualString(op_name));
+      TORCH_INTERNAL_ASSERT(ops.size() == 1);
+      ops_[op_name] = ops.front();
+    }
+
+    return ops_.at(op_name);
+  }
+
+ private:
+  ATenOperatorCache() = default;
+  std::unordered_map<std::string, std::shared_ptr<torch::jit::Operator>> ops_;
+};
+
+// Some arguments of backward operator are not from forward operator's input or output,
+// but need some processing. Need to put such processing code here.
+
+using TensorTransformFunc = std::function<c10::IValue(const at::Tensor&)>;
+
+static const TensorTransformFunc embedding_num_weights = [](const at::Tensor& tensor) {
+  return c10::IValue(tensor.size(0));
+};
+
+static const std::unordered_map<std::string, std::unordered_map<size_t, TensorTransformFunc>> TENSOR_TRANSFORM_FUNCS = {
+    {"aten::embedding_backward", {{2, embedding_num_weights}}},
+};
+
+template <typename T>
+void SetIValueArguments(std::vector<c10::IValue>& ivalue_arguments,
+                        const std::vector<std::tuple<size_t, T>>& raw_arguments) {
+  for (size_t i = 0; i < raw_arguments.size(); i++) {
+    ivalue_arguments[std::get<0>(raw_arguments[i])] = c10::IValue(std::get<1>(raw_arguments[i]));
+  }
+}
+
+// TODO: Add more argument types, such as list type.
+std::vector<DLManagedTensor*> ExecuteATenOperator(
+    const char* op_name, const std::vector<std::tuple<size_t, DLManagedTensor*>>& tensor_arguments,
+    const std::vector<std::tuple<size_t, int64_t>>& int_arguments,
+    const std::vector<std::tuple<size_t, float>>& float_arguments,
+    const std::vector<std::tuple<size_t, bool>>& bool_arguments) {
+  std::string op_name_str(op_name);
+  std::shared_ptr<torch::jit::Operator> op = ATenOperatorCache::Instance().GetOperator(op_name_str);
+  std::vector<c10::IValue> arguments;
+  arguments.resize(op->schema().arguments().size());
```

Comment:
add a todo here, need to handle optional argument and arguments with default values.